File: Coding-Interviews/22-链表中倒数第k个节点.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode* getKthFromEnd(ListNode* head, int k)
    {
        auto cur    = head;
        auto next_k = head;
        for (int i = 0; i < k; i++) {
            next_k = next_k->next;
        }
        while (next_k) {
            next_k = next_k->next;
            cur    = cur->next;
        }
        return cur;
    }
};

File: Coding-Interviews/22-链表中倒数第k个节点.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getKthFromEnd(self, head: ListNode, k: int) -> ListNode:
        cnt = 0
        cur, res = head, head
        while cur and cnt < k:
            cur = cur.next
            cnt += 1
        if not cur:
            return head
        while cur:
            res = res.next
            cur = cur.next
        return res
